### ISSUE TITLE
Tests don't run anymore

### ISSUE THREAD

**Issue by jonsterling** [RedPRL/cooltt#240]
Once again, `make test` does nothing.

**Comment by favonia** [comment]
I believe this means you have run the tests earlier. dune knows that you have not changed the code. You could go into the test directory and run
```
dune exec ./test.exe
```

**Comment by jonsterling** [comment]
Maybe I did run the tests earlier, but I want to run them now! That's what tests are for. LOL.

Anyway, maybe that is a possibility, but the bug is that `make test` and `dune runtest` don't do the same thing every time. So if the `test.exe` thing works, we should change the make task to do that.

**Comment by jonsterling** [comment]
Btw, if I run `./_build/default/test/test.exe` nothing happens!!

**Comment by favonia** [comment]
> Btw, if I run `./_build/default/test/test.exe` nothing happens!!

Did you go into the `test` directory? That was an important step.
```bash
cd test
../_build/default/test/test.exe
```

**Comment by jonsterling** [comment]
I see; that works, but I don't accept it as a solution (too esoteric).

**Comment by favonia** [comment]
> I see; that works, but I don't accept it as a solution (too esoteric).

I see your point, but the "intended" version is much less exotic:
```
cd test
dune exec ./test.exe
```

Theoretically, `dune runtest --force` should trigger the actual testing, but maybe our custom rules overwrote that. @TOTBWF

**Comment by jonsterling** [comment]
I am not willing to be constantly doing this `cd` crap, and I am not a good enough scripter to do the `cd` and have it back-out, while handling errors properly.

But if anyone wants to do that, feel free. I'd also appreciate if anyone could come to my rescue in this thread: https://github.com/ocaml/dune/issues/4390. I feel like I'm a little out of my depth, because I am speaking as an end-user and the other participants are interacting at a much more technical level.

**Comment by jonsterling** [comment]
And it might be that the suggestions in that thread could help us fix our script.

**Comment by ejgallego** [comment]
I continue the discussion here as the thread in the dune repos is off-topic.

> I have to insist that it doesn't really match the way that tests are used by engineers.

Umm, this seems a pretty strong statement given that dune is used for testing codebases in the dozen of millions of ocaml lines, and it was develop in an industrial context. On the contrary, I'd say that it makes little sense to re-run something you know is still valid.

> that test output should never be cached in practice for most use-cases.

What is the point of re-running a pure function when the inputs have not changed? For example, for timing, if you didn't change anything in your repository, the old timings should still be valid. If you changed some dependency, then dune should re-run what is needed.

You can use `(deps universe)` tho, as to indicate, the deps are always out of date if you want, but usually this is not what you want; I agree benchmarking is a bit of an special case tho, as if you don't implement an incremental time aggregation then you want to over-approximate deps or to reset your build graph to a clean state, but caching may interfere too so you may want to disable it in the command line for a timing bench.

**Comment by jonsterling** [comment]
@ejgallego Thanks for dropping by! I don't want to clutter your threads, and I apologize if that was the result of my intervention.

First off, I would just request that you accept that I have a good reason for wanting the tests to really run each time. First of all, it is possible that tests may be non-deterministic (perhaps you would argue that they should not be called "tests", and I would agree, but one still has to have a way to run them even if we agree not to call them "tests"). But let me put that aside, because our tests here are deterministic. I just wanted to point it out.

Regarding timing, it's not at all the case that the old timings would still be valid. Maybe I was playing music on my computer or running a youtube video in the background, and I realized "Oh that probably made my test run slower, let me close that other program and try again". If you don't find this example realistic, I can come up with dozens more... Anyway I don't think it is productive for us to try and exhaust all the possibilities there; instead I would just like to learn what the correct way to reliably force a dune task to be re-run _regardless_ of whether anything has changed. Or if this is something you advise doing outside of dune, then I would also find that confirmation very helpful as well.

Thank you very much!

**Comment by ejgallego** [comment]
My advice is just to think of Dune as a system where rules a pure functions ` { deps } --[action]--> { targets }` , then, add memoization on top of it. The pure nature of rules is essential here as this is what gives incrementality.

For this concrete case, have `make tests` just be a wrapper for an alias `display-tests`, which does `cat` on the tests output.

Then, you let dune handle the test output by itself.

**Comment by ejgallego** [comment]
> But let me put that aside, because our tests here are deterministic. I just wanted to point it out.

Indeed one of the reason dune has the `(universe)` dependency is to handle non-deterministic stuff, so if you have non-deterministic rules you need to add the "universe" dep on it [à la IO monad]

Your point about timing tests is indeed valid, but this would be usually be handled by an special target that setups the right initial state, disables caching, etc... but for regular testing, you gain nothing from re-running.

**Comment by jonsterling** [comment]
thanks! So it seems like maybe I can achieve the behavior I want by using this `universe` thing if I understand you correctly.

Regarding timing, I'm not sure what you mean by 'regular testing', but perhaps we can just put aside the discussion of whether I gain something from re-running. Today I wanted to re-run the tests because I was trying to find out if a particular algorithm change degraded performance or not. I felt that I should just be able to do this with a single command, without doing some complicated profiling target etc. which is more sophisticated than our very humble-at-the-moment needs.

**Comment by jonsterling** [comment]
resolved by #241
